### PR TITLE
Wasm: implement Services API

### PR DESCRIPTION
**Pull request by mykmartin**
This limits request/response objects to key:value string pairs.

@alxrsngrtn API changes: added serviceRequest import, serviceResponse export, Particle.init method.

### CODE REVIEW

File: src/runtime/particle.ts [hunk 1]
```diff
@@ -20,6 +20,11 @@ import {Content} from './slot-consumer.js';
 import {UserException} from './arc-exceptions.js';
 import {Entity, EntityRawData, MutableEntityData} from './entity.js';
 
+export interface Capabilities {
+  constructInnerArc?: (particle: Particle) => Promise<InnerArcHandle>;
+  serviceRequest?: (particle: Particle, args, callback) => void;
```
Comment (alxrsngrtn) [inline src/runtime/particle.ts]: I like this abstraction. 
I feel like it will make #3245 easier to implement.

File: src/runtime/wasm.ts [hunk 2]
```diff
@@ -31,6 +31,9 @@ import {Dictionary} from './hot.js';
 // Examples:
 //   Singleton:   4:id05|txt:T3:abc|lnk:U10:http://def|num:N37:|flg:B1|
 //   Collection:  3:29:4:id12|txt:T4:qwer|num:N9.2:|18:6:id2670|num:N-7:|15:5:id501|flg:B0|
+//
+// The encoder classes also support a "Dictionary" format of key:value string pairs:
+//   <size>:<key-len>:<key><value-len>:<value><key-len>:<key><value-len>:<value>...
```
Comment (alxrsngrtn) [inline src/runtime/wasm.ts]: Can we format this string as a jsdoc? i.e. `/** ... */`?

Comment (alxrsngrtn) [reply src/runtime/wasm.ts]: It will help the generation of https://live.arcs.dev/dist/apidocs/

File: src/runtime/wasm.ts [hunk 3]
```diff
@@ -422,6 +434,11 @@ export class WasmContainer {
     return p;
   }
 
+  // Convenience function for freeing one or more wasm memory allocations. Null pointers are ignored.
```
Comment (alxrsngrtn) [inline src/runtime/wasm.ts]: jsdoc

File: src/runtime/wasm.ts [hunk 4]
```diff
@@ -422,6 +434,11 @@ export class WasmContainer {
...
+  free(...ptrs: WasmAddress[]) {
+    ptrs.forEach(p => p && this.exports._free(p));
```
Comment (alxrsngrtn) [inline src/runtime/wasm.ts]: :+1:

File: src/runtime/wasm.ts [hunk 5]
```diff
@@ -651,11 +668,42 @@ export class WasmParticle extends Particle {
     }
   }
 
+  // Wasm particles can request service calls with a Dictionary of arguments and an optional string
+  // tag to disambiguate different requests to the same service call.
+  async serviceRequest(callPtr: WasmAddress, argsPtr: WasmAddress, tagPtr: WasmAddress) {
+    const call = this.container.read(callPtr);
+    const args = new StringDecoder().decodeDictionary(this.container.read(argsPtr));
+    const tag = this.container.read(tagPtr);
```
Comment (alxrsngrtn) [inline src/runtime/wasm.ts]: Can we generate a default value for the tag in case it isn't supplied?

File: src/wasm/cpp/arcs.h [hunk 6]
```diff
@@ -603,27 +633,38 @@ class Particle {
...
   virtual void onHandleUpdate(Handle* handle) {}
+
+  // Override to react to UI events triggered by handlers in the template provided below.
   virtual void fireEvent(const std::string& slot_name, const std::string& handler) {}
 
   // Override to provide a template string and key:value model for rendering into a slot.
   virtual std::string getTemplate(const std::string& slot_name) { return ""; }
-  virtual void populateModel(const std::string& slot_name,
-                             std::function<void(const std::string&, const std::string&)> add) {}
+  virtual void populateModel(const std::string& slot_name, Dictionary* model) {}
```
Comment (alxrsngrtn) [inline src/wasm/cpp/arcs.h]: :+1: I think a dictionary is a better interface than a callback function.

File: src/wasm/cpp/arcs.h [hunk 7]
```diff
@@ -572,27 +597,32 @@ class Particle {
...
+  // Convenience method for sub-classes to retrieve a handle by name.
   Handle* getHandle(const std::string& name) const {
     auto it = handles_.find(name);
     return (it != handles_.end()) ? it->second : nullptr;
   }
 
-  void renderSlot(const std::string& slot_name, bool send_template, bool send_model) {
+  // Can be called by sub-classes to initiate rendering; also invoked when auto-render is enabled
+  // after all handles have been synchronized.
+  void renderSlot(const std::string& slot_name, bool send_template = true, bool send_model = true) {
```
Comment (alxrsngrtn) [inline src/wasm/cpp/arcs.h]: :+1: default arguments.

File: src/wasm/cpp/test-particle.cc [hunk 8]
```diff
@@ -225,5 +225,54 @@ class SimpleParticle : public arcs::Particle {
   arcs::Info local_;
 };
 
-DEFINE_PARTICLE(TestParticle)
 DEFINE_PARTICLE(SimpleParticle)
+
+
+class ServiceParticle : public arcs::Particle {
```
Comment (alxrsngrtn) [inline src/wasm/cpp/test-particle.cc]: This is a really cool example. Thanks for choosing to convert this service area to cpp wasm.

File: src/runtime/wasm.ts (hunk 2 above)
Comment (mykmartin) [reply src/runtime/wasm.ts]: I'll do that in a separate CL, to avoid muddying the logical changes here.

File: src/runtime/wasm.ts (hunk 5 above)
Comment (mykmartin) [reply src/runtime/wasm.ts]: That's already done on the C++ side.

However, this code should be robust against incorrect usage (either accidental or malicious), so all of the wasm-to-js calls should have better parameters guards. I'll add a TODO.

File: src/runtime/wasm.ts (hunk 2 above)
Comment (mykmartin) [reply src/runtime/wasm.ts]: Actually, not really sure of the utility for that. Having generated docs for internal classes doesn't seem all that useful - I certainly just look at the actual code for details on APIs.

**Pull request merged**